src/ophys_etl/modules/denoising/fine_tuning/utils/data_splitter.py

```diff
@@ -0,0 +1,64 @@
+from pathlib import Path
+from typing import Union, Tuple, List, Optional
+
+import h5py
+import numpy as np
+
+
+class DataSplitter:
+    """Splits movie into train/test sets of frames"""
+    def __init__(
+        self,
+        movie_path: Union[str, Path],
+        seed: Optional[int] = None
+    ):
+        """
+
+        Parameters
+        ----------
+        movie_path
+            Path to movie
+        seed
+            Seed for reproducibility
+        """
+        self._movie_path = Path(movie_path)
+        self._seed = seed
+
+    def get_train_val_split(
+            self,
+            train_frac: float = 0.7,
+            window_size: int = 30
+    ) -> Tuple[List[int], List[int]]:
+        """
+        Returns train/val split for the movie. Each split contains the center
+        frame indices.
+
+        Parameters
+        ----------
+        train_frac
+            The fraction of center frames to reserve for the training set
+        window_size
+            Number of frames before and after a center frame
+
+        Returns
+        -------
+        Tuple of train, val center frame indices
+        """
+        with h5py.File(self._movie_path, 'r') as f:
+            nframes = f['data'].shape[0]
+
+        # min, max center frame must be window_size away from beginning/end
+        # since the frame needs to have window_size before/after it
+        all_frames = np.arange(window_size,
+                               nframes - window_size)
+
+        # Reserving the last (1 - train_frac) frames for validation
+        n_train = int(len(all_frames) * train_frac)
+
+        train = all_frames[:n_train]
+        val = all_frames[n_train:]
```

Comment:
Are we worried about this kind of split at all? Looking at Ophys movies, I've seen a lot that have very easy to spot trends in noise etc. over the course of the movie. I assume this isn't that large of an issue as any info not in the train will be in the validation and it can kind of learn it? Dunno, wonder if we could end up with movies with poor validation stats.

Reply:
I've thought a lot about this. If we take the last 30% of the movie for validation, I think it is as new and different from the first 70% as possible. It is also the easiest approach.

There are a couple alternatives I've thought of:
1) Take all chunks of the movie with stride 1, (i.e. frames 1-60, frames 2-61, 3 - 62, ...), shuffle this, then split into train, validation. Let's say the chunk of 1-60 ended up in train and 2-61 in validation. Both of these chunks of the movie are obviously highly correlated, and I think how the model performs on the validation set would not be considered "generalization" performance

2) To help with the above issue, we could try to use a stride of the window size, i.e frames 1-60, 61-121, ...) then shuffle and split into train/val. The problem here is that it significantly reduces the amount of training data. I think for ~20,000 frame movie, this results in ~500 training examples. 

3) As middle ground, we could make sure the validation frames are more than x frames away from the training frames. This adds complexity. And this still has the issue that the frames in train/val would overlap and would be correlated. I.e we might have frames 1-60 in train, but frames 10-70 in validation, which still has a lot of overlap.

I should also mention that I thought about this and couldn't really convince myself that these metrics on the validation set are even meaningful. The purpose here is to measure "generalization" on unseen data. The typical use case would be that you train on data, deploy model, then evaluate it on a bunch of data. You want to know how well your model will do on the unseen data. Here though, we are training on a single movie, then deploying it on that movie. So we don't have to worry about unseen data. So shouldn't we only care about how well the model can learn the signal from the movie? I.e. how much it can reduce the loss function on the movie. 

In the end my thinking is that it's still somewhat useful as it gives us some measure of how well did it learn the signal on the movie by testing it on new data from the same movie. But I am not sure how useful it is.

Reply:
Yeah, I'm remembering this conversation. Hmm... yeah, the options you list above aren't sparking joy. Options I've come up with are:

- Pull the validation data from the center or ends of the movie. If you assume any trend is smooth over the course of the movie, pulling from the center would train the data on the extrema and see if the model can extrapolate between. Pulling validation from the ends (%val/2 on each end) would kind of do the same thing but flipped.
- We could also chunk the data into halves/quarters/etc (making it a configurable) and pull the percentages for train/val from each chunk.

None of these are perfect or I can really rationalize over any other choice. Especially in light of how meaningful you think the validation metrics are.

Reply:
Let's talk about this.

Reply:
let's keep it as is as we discussed.